### ISSUE TITLE
EF Core does not populate subcollections in foreign key relationships in many to many tables

### ISSUE THREAD

**Issue by KaiWesterschwiensterdt** [dotnet/efcore#27149]
## File a bug

Authors are to be assigned to books. A book can have several authors, several authors can have written several books. 
First the book is to be created, then the authors are to be read out and then the link between the two is to be written in a link table.

The bug:
When adding the author_books to the databasecontext, only the subcollections of the books are filled correctly, but the subcollection of the authors is assigned as null.

For example a generated authorBook:
```C#
    {
       ID: 1,
       BookID: 152,
       AutherID: 9

       Book: {bookObject},                
       Author: null
    }
```

### Include your code

Models:
```C#
public class Book
        {
            [Key]
            public int ID { get; set; }
            public int Name{ get; set; }

            [InverseProperty("Author")]
            public virtual ICollection<Author_Book> Author_Book{ get; set; }
        }

public class Author
        {
            [Key]
            public int ID { get; set; }
            public int Name{ get; set; }

            [InverseProperty("Book")]
            public virtual ICollection<Author_Book> Author_Book{ get; set; }
        }

public class Author_Book
        {
            [Key]
            public int ID { get; set; }
            public int BookID { get; set; }
            public int AuthorID { get; set; }
    
            [ForeignKey(nameof(BookID ))]
            [InverseProperty(nameof(Book.Author_Book))]
            public virtual Book Book{ get; set; }
            [ForeignKey(nameof(AuthorID ))]
            [InverseProperty(nameof(Author.Author_Book))]
            public virtual Author Author{ get; set; }
        }
```

Logic:
```C#
public async void Create()
        {
            try
            {
                ICollection<Author_Book> authorBooks = new List<Author_Book>();
                Book book= new Book
                {
                    Name = "test"
                };
                book = await _bookRepository.Add(book);
                List<Author> authors = await _authorRepository.Get();

                foreach (Author author in authors )
                {
                    Author_Book authorBook = new Author_Book 
                    {
                        BookID = book.ID,
                        Author ID = author .ID,
                    };

                    authorBooks .Add(authorBook);
                }
                authorBooks = await _authorBookRepository.Add(authorBooks );
        }
```

EF Logic:
```C#
        public async Task<Book> Add(Book book)
        {
                _databaseContext.Books
                .Add(book);

            await this._databaseContext.SaveChangesAsync();

            return book;
        }

        public async Task<ICollection<Author>> Get()
        {
            return await _databaseContext.Author
                .OrderBy(x => x.ID)
                .ToListAsync();
        }

_________________________________________________Exception_here__________________________________________________
public async Task<ICollection<Author_Book>> CreateAsync(ICollection<Author_Book> authorsBooks)
        {
            int i = 0;
            foreach (var autherBook in authorsBooks)
            {
                _databaseContext.Author_Book
                .Add(authorBook);
                i++;
    
                if (i == 99)
                {
                    await this._databaseContext.SaveChangesAsync();
                    i = 0;
                }
            }
            // TODO: bugfix -> ID always 0, error output "Can not insert value NULL"
            await this._databaseContext.SaveChangesAsync();
    
            return authorsBooks;
    
        }
_________________________________________________Exception_here__________________________________________________
```

### Include stack traces

Include the full exception message and stack trace for any exception you encounter.

Use triple-tick fences for stack traces. For example:

```
Microsoft.Data.SqlClient.SqlException (0x80131904): Cannot insert the value NULL into column 'ID', table 'Library_Test.dbo.Author_Book'; column does not allow nulls. UPDATE fails.
      The statement has been terminated.
         at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__188_0(Task`1 result)
         at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
         at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
         at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
      --- End of stack trace from previous location ---
         at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
         at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
      --- End of stack trace from previous location ---
         at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
      ClientConnectionId:3d198da5-3602-4490-a8ac-74af2cf73c53
      Error Number:515,State:2,Class:16
```

### Include provider and version information

EF Core version: 6.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: NET 6.0
Operating system: Windows 10
IDE:Visual Studio 2022 17.0.4

![ProgramFlowChart](https://user-images.githubusercontent.com/61865463/148763914-854205f0-8967-4029-b520-826be3734064.jpg)

**Comment by KaiWesterschwiensterdt** [comment]
![ProgramFlowChart](https://user-images.githubusercontent.com/61865463/148763861-0fd0d24e-42c5-4034-ac0b-339ec93db3f6.jpg)

**Comment by ajcvickers** [comment]
@KaiWesterschwiensterdt This issue is lacking enough information for us to be able to fully understand what is happening. Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by KaiWesterschwiensterdt** [comment]
I have read your comment and today directly wrote a small program. For this I had to change a few settings on the database. When I then tested my small, finished program, everything ran flawlessly. Then it dawned on me. So I tested my real program and there it worked. There were settings in the database that were not optimal. The properties for changing database columns was disabled. The link table Author_Book had an error with the identity columns. Without your request for a small program I probably would not have found it for a long time. Many thanks for that. #herooftheday
